In [1]:
# Import the modules
import pandas as pd
import numpy as np
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report
from sklearn.svm import SVC
import tensorflow as tf
import ydf
from sklearn.naive_bayes import GaussianNB
from xgboost import XGBClassifier

## Dataset 2 - use one hot encoding with StandardScaler to deal with dataset

In [2]:
# Read in the cleaned/prepared csv file
# c:\Users\schre\Downloads\diabetes_prediction_cleaned_sn.csv
dm_prediction_df = pd.read_csv("Resources/diabetes_prediction_cleaned_one_hot_encoding.csv")
dm_prediction_df.head()

age  hypertension  heart_disease       bmi  hba1c_level  \
0  1.691761             0              1 -0.282234     0.787897   
1  0.538015             0              0  0.018883     0.787897   
2 -0.615730             0              0  0.018883    -0.120279   
3 -0.260731             0              0 -0.583350    -0.120279   
4  1.514261             1              1 -1.035025    -1.028455   

   blood_glucose_level  diabetes_status  gender_Female  gender_Male  \
0             0.047704                0              1            0   
1            -1.426210                0              1            0   
2             0.489878                0              0            1   
3             0.416183                0              1            0   
4             0.416183                0              0            1   

   gender_Other  smoking_history_No Info  smoking_history_current  \
0             0                        0                        0   
1             0                        1                        0   
2             0                        0                        0   
3             0                        0                        1   
4             0                        0                        1   

   smoking_history_ever  smoking_history_former  smoking_history_never  \
0                     0                       0                      1   
1                     0                       0                      0   
2                     0                       0                      1   
3                     0                       0                      0   
4                     0                       0                      0   

   smoking_history_not current  
0                            0  
1                            0  
2                            0  
3                            0  
4                            0

# Create the labels set (y) from the “diabetes_status” column, and then create the features (X) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features
# Separate the y variable, the labels
y = dm_prediction_df['diabetes_status']

# Separate the X variable, the features
X = dm_prediction_df.drop(columns=['diabetes_status'])

In [4]:
# Review the y variable Series
display(y.head())
display(y.tail())

0    0
1    0
2    0
3    0
4    0
Name: diabetes_status, dtype: int64

99995    0
99996    0
99997    0
99998    0
99999    0
Name: diabetes_status, dtype: int64

In [5]:
# Review the X variable DataFrame
display(X.head())
display(X.tail())

age  hypertension  heart_disease       bmi  hba1c_level  \
0  1.691761             0              1 -0.282234     0.787897   
1  0.538015             0              0  0.018883     0.787897   
2 -0.615730             0              0  0.018883    -0.120279   
3 -0.260731             0              0 -0.583350    -0.120279   
4  1.514261             1              1 -1.035025    -1.028455   

   blood_glucose_level  gender_Female  gender_Male  gender_Other  \
0             0.047704              1            0             0   
1            -1.426210              1            0             0   
2             0.489878              0            1             0   
3             0.416183              1            0             0   
4             0.416183              0            1             0   

   smoking_history_No Info  smoking_history_current  smoking_history_ever  \
0                        0                        0                     0   
1                        1                        0                     0   
2                        0                        0                     0   
3                        0                        1                     0   
4                        0                        1                     0   

   smoking_history_former  smoking_history_never  smoking_history_not current  
0                       0                      1                            0  
1                       0                      0                            0  
2                       0                      1                            0  
3                       0                      0                            0  
4                       0                      0                            0

age  hypertension  heart_disease       bmi  hba1c_level  \
99995  1.691761             0              0  0.018883     0.787897   
99996 -1.769475             0              0 -1.486700     0.787897   
99997  1.070513             0              0  0.018883    -0.120279   
99998 -0.793229             0              0  1.223350    -1.028455   
99999  0.671140             0              0 -0.733909     0.787897   

       blood_glucose_level  gender_Female  gender_Male  gender_Other  \
99995            -1.180558              1            0             0   
99996            -0.934905              1            0             0   
99997             0.416183              0            1             0   
99998            -0.934905              1            0             0   
99999            -1.180558              1            0             0   

       smoking_history_No Info  smoking_history_current  smoking_history_ever  \
99995                        1                        0                     0   
99996                        1                        0                     0   
99997                        0                        0                     0   
99998                        0                        0                     0   
99999                        0                        1                     0   

       smoking_history_former  smoking_history_never  \
99995                       0                      0   
99996                       0                      0   
99997                       1                      0   
99998                       0                      1   
99999                       0                      0   

       smoking_history_not current  
99995                            0  
99996                            0  
99997                            0  
99998                            0  
99999                            0

In [6]:
print(X.shape)

(100000, 15)


## Split the data into training and testing datasets by using train_test_split.

In [7]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split. Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    random_state=1, 
                                                    stratify=y)

# Create a Support Vector Machine Model with the Original Data
## Fit a Support Vector Machine model by using the training data (X_train and y_train).

In [8]:
model = SVC(kernel='linear')
model.fit(X_train, y_train)

SVC(kernel='linear')

## Make predictions on the test set and compute accuracy / classification report

In [9]:
predictions = model.predict(X_test)
predictions

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [10]:
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           0       0.96      1.00      0.98     22875
           1       0.93      0.56      0.70      2125

    accuracy                           0.96     25000
   macro avg       0.95      0.78      0.84     25000
weighted avg       0.96      0.96      0.95     25000



## Try a nonlinear kernel for the support vector classifier

In [11]:
model = SVC(kernel='poly')
model.fit(X_train, y_train)

predictions = model.predict(X_test)
predictions

print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           0       0.97      1.00      0.98     22875
           1       1.00      0.62      0.76      2125

    accuracy                           0.97     25000
   macro avg       0.98      0.81      0.87     25000
weighted avg       0.97      0.97      0.96     25000



## Neural Network Models

In [12]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = 15
hidden_nodes_layer1 =  30
hidden_nodes_layer2 = 5

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
)

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()


Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 30)                480       
                                                                 
 dense_1 (Dense)             (None, 5)                 155       
                                                                 
 dense_2 (Dense)             (None, 1)                 6         
                                                                 
Total params: 641 (2.50 KB)
Trainable params: 641 (2.50 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [13]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [14]:
# Train the model
fit_model = nn.fit(X_train,y_train,epochs=100)

Epoch 1/100


2344/2344 [==============================] - 2s 569us/step - loss: 0.1427 - accuracy: 0.9489
Epoch 2/100
2344/2344 [==============================] - 1s 560us/step - loss: 0.1098 - accuracy: 0.9649
Epoch 3/100
2344/2344 [==============================] - 1s 564us/step - loss: 0.1004 - accuracy: 0.9667
Epoch 4/100
2344/2344 [==============================] - 1s 565us/step - loss: 0.0954 - accuracy: 0.9677
Epoch 5/100
2344/2344 [==============================] - 1s 564us/step - loss: 0.0930 - accuracy: 0.9677
Epoch 6/100
2344/2344 [==============================] - 1s 555us/step - loss: 0.0920 - accuracy: 0.9681
Epoch 7/100
2344/2344 [==============================] - 1s 552us/step - loss: 0.0918 - accuracy: 0.9682
Epoch 8/100
2344/2344 [==============================] - 1s 558us/step - loss: 0.0914 - accuracy: 0.9680
Epoch 9/100
2344/2344 [==============================] - 1s 553us/step - loss: 0.0912 - accuracy: 0.9683
Epoch 10/100
2344/2344 [=============================

In [15]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

782/782 - 0s - loss: 0.0941 - accuracy: 0.9675 - 380ms/epoch - 486us/step
Loss: 0.09405865520238876, Accuracy: 0.967519998550415


## Step 2 - using keras tuner to select the highest performing model

In [16]:
# Create a method that creates a new Sequential model with hyperparameter options
def create_model(hp):
    nn_model = tf.keras.models.Sequential()

    # Allow kerastuner to decide which activation function to use in hidden layers
    activation = hp.Choice('activation',['relu','tanh','sigmoid'])
    
    # Allow kerastuner to decide number of neurons in first layer
    nn_model.add(tf.keras.layers.Dense(units=hp.Int('first_units',
        min_value=1,
        max_value=30,
        step=2), activation=activation, input_dim=15))

    # Allow kerastuner to decide number of hidden layers and neurons in hidden layers
    for i in range(hp.Int('num_layers', 1, 6)):
        nn_model.add(tf.keras.layers.Dense(units=hp.Int('units_' + str(i),
            min_value=1,
            max_value=10,
            step=2),
            activation=activation))
    
    nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

    # Compile the model
    nn_model.compile(loss="binary_crossentropy", optimizer='adam', metrics=["accuracy"])
    
    return nn_model

In [17]:
# Import the kerastuner library
import keras_tuner as kt

tuner = kt.Hyperband(
    create_model,
    objective="val_accuracy",
    max_epochs=20,
    hyperband_iterations=2)

In [18]:
# Run the kerastuner search for best hyperparameters
tuner.search(X_train,y_train,epochs=20,validation_data=(X_test,y_test))

Trial 60 Complete [00h 00m 44s]
val_accuracy: 0.9682000279426575

Best val_accuracy So Far: 0.968280017375946
Total elapsed time: 00h 15m 56s


In [19]:
# Get best model hyperparameters
best_hyper = tuner.get_best_hyperparameters(1)[0]
best_hyper.values

{'activation': 'relu',
 'first_units': 15,
 'num_layers': 2,
 'units_0': 5,
 'units_1': 5,
 'units_2': 9,
 'units_3': 7,
 'units_4': 3,
 'units_5': 9,
 'tuner/epochs': 20,
 'tuner/initial_epoch': 7,
 'tuner/bracket': 2,
 'tuner/round': 2,
 'tuner/trial_id': '0045'}

In [20]:
# Evaluate best model against full test data
best_model = tuner.get_best_models(1)[0]
model_loss, model_accuracy = best_model.evaluate(X_test,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

782/782 - 0s - loss: 0.0929 - accuracy: 0.9683 - 430ms/epoch - 549us/step
Loss: 0.09293306618928909, Accuracy: 0.968280017375946


## Yggdrasil decision tree model

The instructors suggested using the Yggdrasil decision tree model for a machine learning classification problem, so we decided to try it out and see if it could improve model performance. The accuracy was 0.967, which is not quite as good as svm, so we did not include it in the final analysis.

In [21]:
#need X_train, y_train combined into one dataframe
train_df=pd.concat([X_train,y_train],axis=1)
train_df.head()

test_df=pd.concat([X_test,y_test],axis=1)
test_df.head()

age  hypertension  heart_disease       bmi  hba1c_level  \
39679 -0.837604             0              0 -0.131675    -1.028455   
56089 -1.547601             0              0 -1.637259    -1.028455   
86698  1.070513             0              0 -0.131675    -1.028455   
64026 -1.769475             0              0 -2.088934     0.787897   
33740 -0.216357             0              0  4.234517    -1.028455   

       blood_glucose_level  gender_Female  gender_Male  gender_Other  \
39679             0.539009              0            1             0   
56089             0.539009              0            1             0   
86698             0.514444              0            1             0   
64026            -1.426210              1            0             0   
33740             1.521618              0            1             0   

       smoking_history_No Info  smoking_history_current  smoking_history_ever  \
39679                        0                        0                     0   
56089                        1                        0                     0   
86698                        1                        0                     0   
64026                        1                        0                     0   
33740                        0                        0                     1   

       smoking_history_former  smoking_history_never  \
39679                       0                      1   
56089                       0                      0   
86698                       0                      0   
64026                       0                      0   
33740                       0                      0   

       smoking_history_not current  diabetes_status  
39679                            0                0  
56089                            0                0  
86698                            0                0  
64026                            0                0  
33740                            0                0

In [22]:
model = ydf.RandomForestLearner(label='diabetes_status',
                                task=ydf.Task.CLASSIFICATION).train(train_df)

Train model on 75000 examples
Model trained in 0:00:01.385456


In [23]:
evaluation = model.evaluate(test_df)

print(evaluation)

accuracy: 0.96784
confusion matrix:
    label (row) \ prediction (col)
    +-------+-------+-------+
    |       |     0 |     1 |
    +-------+-------+-------+
    |     0 | 22862 |    13 |
    +-------+-------+-------+
    |     1 |   791 |  1334 |
    +-------+-------+-------+
characteristics:
    name: '1' vs others
    ROC AUC: 0.931697
    PR AUC: 0.834077
    Num thresholds: 182
loss: 0.36414
num examples: 25000
num examples (weighted): 25000



## Gaussian Naive Bayes

Gaussian Naive Bayes is a machine learning classification technique based on a probablistic approach that assumes each class follows a normal distribution. It assumes each parameter has an independent capacity of predicting the output variable. The model makes strong assumptions about independence of the predictors, which our data violates, so the model did not end up with a very high accuracy and was excluded from our final analysis.

In [24]:
gnb = GaussianNB()
predictions = gnb.fit(X_train, y_train).predict(X_test)

In [25]:
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           0       0.99      0.61      0.75     22875
           1       0.18      0.96      0.31      2125

    accuracy                           0.64     25000
   macro avg       0.59      0.78      0.53     25000
weighted avg       0.92      0.64      0.71     25000



## XGBoost model

XGBoost is a gradient boosted decision tree method. It is a regularized version of the gradient boosting algorithm, which makes it faster and more effective, especially when datasets include numerical and categorical features. It has historically been used to great effect in machine learning competitions such as on the Kaggle platform. It produced great results for us with 97 percent accuracy.

In [26]:
bst = XGBClassifier(n_estimators=2, max_depth=2, learning_rate=1, objective='binary:logistic')
# fit model
bst.fit(X_train, y_train)
# make predictions
predictions = bst.predict(X_test)

In [27]:
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           0       0.97      1.00      0.98     22875
           1       1.00      0.62      0.76      2125

    accuracy                           0.97     25000
   macro avg       0.98      0.81      0.87     25000
weighted avg       0.97      0.97      0.96     25000

